src/auspex/instruments/prologix.py

```diff
@@ -0,0 +1,292 @@
+
+import os
+import numpy as np
+import socket
+import functools
+from auspex.log import logger
+
+_converters = {
+    's': str,
+    'b': functools.partial(int, base=2),
+    'c': chr,
+    'd': int,
+    'o': functools.partial(int, base=8),
+    'x': functools.partial(int, base=16),
+    'X': functools.partial(int, base=16),
+    'e': float,
+    'E': float,
+    'f': float,
+    'F': float,
+    'g': float,
+    'G': float,
+}
+
+class PrologixError(Exception):
+    """Error interacting with the Prologix GPIB-ETHERNET controller."""
+
+class PrologixSocketResource(object):
+    """A resource representing a GPIB instrument controlled through a PrologixError
+    GPIB-ETHERNET controller. Mimics the functionality of a pyVISA resource object.
+
+    See http://prologix.biz/gpib-ethernet-controller.html for more details
+    and a utility that will discover all prologix instruments on the network.
+
+    Attributes:
+        timeout: Timeout duration for TCP comms. Default 5s.
+        write_termination: Character added to each outgoing message.
+        read_termination: Character which denotes the end of a reponse message.
+        idn_string: GPIB identification command. Defaults to '*IDN?'
+        bufsize: Maximum amount of data to be received in one call, in bytes.
+    """
+    def __init__(self, ipaddr=None, gpib=None):
+        super(PrologixSocketResource, self).__init__()
+        if ipaddr is not None:
+            self.ipaddr = ipaddr
+        if gpib is not None:
+            self.gpib = gpib
+        self.sock = None
+        self._timeout = 5
+        self.read_termination = "\r\n"
+        self.write_termination = "\r\n"
+        self.idn_string = "*IDN?"
+        self.bufsize = 4096
+
+    @property
+    def timeout(self):
+        return self._timeout
+
+    @timeout.setter
+    def timeout(self, value):
+        self._timeout = timeout
+        if self.sock is not None:
+            self.sock.settimeout(self._timeout)
+
+    def connect(self, ipaddr=None, gpib=None):
+        """Connect to a GPIB device through a Prologix GPIB-ETHERNET controller.
+        box.
+
+        Args:
+            ipaddr: The IP address of the Prologix GPIB-ETHERNET.
+            gpib: The GPIB address of the instrument to be controlled.
+        Returns:
+            None.
+        """
+        if ipaddr is not None:
+            self.ipaddr = ipaddr
+        if gpib is not None:
+            self.gpib = gpib
+        try:
+            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM,
+                socket.IPPROTO_TCP)
+            self.sock.settimeout(self._timeout)
+            self.sock.connect((self.ipaddr, 1234)) #Prologix communicates on port 1234
+        except socket.error as err:
+            logger.error("Cannot open socket to Prologix at {0}: {1}".format(self.ipaddr,
+                err.msg))
+            raise PrologixError(self.ipaddr) from err
+        self.sock.send("++ver\r\n".encode())
```

Comment:
for all these constant strings it would be a bit cleaner to use the b prefix as in `self.sock.send(b"++ver\r\n")